system-helper/flatpak-system-helper.c

```diff
@@ -1413,22 +1424,21 @@ static void
 name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
 {
   const gchar *unique_name = (const gchar *) user_data;
-  OngoingPull *pull;
-  g_autoptr(GPtrArray) remove_pulls = NULL;
+  g_autoptr(GPtrArray) cleanup_pulls = NULL;
+  GHashTableIter iter;
+  gpointer value;
 
-  remove_pulls = g_ptr_array_new_with_free_func ((GDestroyNotify) ongoing_pull_free);
+  cleanup_pulls = g_ptr_array_new_with_free_func ((GDestroyNotify) ongoing_pull_free);
 
-  G_LOCK (ongoing_pulls);
-  for (guint i = 0; i < ongoing_pulls->len; i++)
+  G_LOCK (cache_dirs_in_use);
+  g_hash_table_iter_init (&iter, cache_dirs_in_use);
+  while (g_hash_table_iter_next (&iter, NULL, &value))
     {
-      pull = g_ptr_array_index (ongoing_pulls, i);
+      OngoingPull *pull = (OngoingPull *) value;
       if (g_strcmp0 (pull->unique_name, unique_name) == 0)
-        {
-          g_ptr_array_remove (ongoing_pulls, pull);
-          g_ptr_array_add (remove_pulls, pull);
-        }
+        g_ptr_array_add (cleanup_pulls, pull);
```

Comment:
Surely you also want `g_hash_table_iter_remove()` to remove it from the hash table?

Reply:
The `cleanup_pulls` has `ongoing_pull_free` as `GDestroyNotify`.
The `ongoing_pull_free` takes care of removing it from the `hashtable.

Does it makes sense?

Reply:
No, consider the following:

1. Thread A reaches this code path, claims the lock, puts pull P into cleanup_pulls, and releases the lock
2. Thread A yields to thread B
3. Thread B calls `take_ongoing_pull_by_dir` which removes P from the hash table and returns it

Now both A and B will try to free P.

Reply:
Review accepted and addressed in the latest round of iteration.